ramips: add support for Creality WB-01

ihrapsa opened the pull request:
I guess this could be merged with master

adschm commented on target/linux/generic/pending-5.4/483-mtd-spi-nor-add-support-for-Boyamicro-BY25Q128AS.patch:
```diff
@@ -0,0 +1,26 @@
+Subject: [PATCH] mtd: spi-nor: add support for Boya BY25Q128AS
+
+Adds support for Boya Microelectronics (BoyaMicro) BY25Q128AS 128 MiB flash.
+Tested on Creality WB-01 embedded device which uses this chip.
```
Is this submitted upstream, i.e. correctly placed in the pending-* directory?

figgyc replied:
Not currently, apologies. Would you rather I move it to the correct folder or submit it to upstream? (also working on the other changes, just need to do some testing for those so it'll be a bit)

adschm replied:
Submitting upstream is obviously the desired option, as this will have the proper authority care about the change.

adschm commented on target/linux/ramips/dts/mt7628an_creality_wb-01.dts:
```diff
@@ -0,0 +1,116 @@
+		led-failsafe = &led_status;
+		led-upgrade = &led_status;
+		led-running = &led_status;
+		label-mac-device = &wmac;
```
This only works in combination with mtd-mac-address. You will need to set the label MAC address in the lower section of 02_network; the correct lines are already there for vocore,vocore2 ...

adschm commented on target/linux/ramips/dts/mt7628an_creality_wb-01.dts:
```diff
@@ -0,0 +1,116 @@
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
Can this go faster?

adschm commented on target/linux/ramips/dts/mt7628an_creality_wb-01.dts:
```diff
@@ -0,0 +1,116 @@
+&state_default {
+	gpio {
+		groups = "i2s", "refclk", "wdt", "p4led_an",
+				"p3led_an", "p2led_an", "p1led_an", "wled_an";
```
I end up with just `"wdt", "p3led_an", "p1led_an", "p0led_an", "wled_an"` ?

adschm commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -13,6 +13,27 @@ define Build/ravpower-wd009-factory
 	@mv $@.new $@
 endef
 
+define Build/creality_wb-01-factory
```
Please move this up so Build/* recipes are sorting alphabetically.

adschm commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -13,6 +13,27 @@ define Build/ravpower-wd009-factory
 endef
 
+define Build/creality_wb-01-factory
+  echo $@
```
What's that supposed to do?

adschm commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -13,6 +13,27 @@ define Build/ravpower-wd009-factory
+  echo '#!/bin/sh' > $@.install.sh
+  echo 'kernel_size=851968' >> $@.install.sh
+  echo 'file=$$2/factory.bin' >> $@.install.sh
+  echo 'file_size=`wc -c < $$file`' >> $@.install.sh
```
Deprecated backticks might be replaced by $() here ...

adschm commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -13,6 +13,27 @@ define Build/ravpower-wd009-factory
+  echo 'kernel_size=851968' >> $@.install.sh
+  echo 'file=$$2/factory.bin' >> $@.install.sh
+  echo 'file_size=`wc -c < $$file`' >> $@.install.sh
+  echo 'fs_size=`expr $$file_size - $$kernel_size`' >> $@.install.sh
```
Why not simply `fs_size=$(($$file_size - $$kernel_size))` ?

adschm commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -63,6 +84,17 @@ define Device/buffalo_wcr-1166ds
 
+define Device/creality_wb-01
+  IMAGE_SIZE := 16064k
+  IMAGES += cxsw_update.tar.bz2
```
I'd still call this factory.bin, or at least factory.tar.bz2.

The name of the images should indicate the use from an OpenWrt perspective, i.e. factory means it's for initial flashing.
Since renaming the image will be necessary anyway, one can also rename from factory.* to the proper name then.

adschm commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -63,6 +84,17 @@ define Device/buffalo_wcr-1166ds
+  DEVICE_VENDOR := Creality
+  DEVICE_MODEL := WB-01
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-ohci kmod-sdhci-mt7620
+  SUPPORTED_DEVICES += creality_wb-01
```
This is default and can be dropped.

adschm replied:
Now it's not the default anymore, but just wrong. Please drop SUPPORTED_DEVICES.

adschm commented on target/linux/ramips/mt76x8/base-files/etc/board.d/01_leds:
```diff
@@ -20,6 +20,10 @@ netgear,r6120)
 	;;
+creality,wb-01)
+	ucidef_set_led_switch "lan" "lan" "blue:lan" "switch0" "0x04"
+	ucidef_set_led_netdev "wlan2g" "wlan2g" "yellow:wlan2g" "phy0tpt"
```
WiFi trigger is already in the DTS and can be dropped here.

adschm commented:
Consider renaming your PR to something like "ramips: add support for Creality WB-01" as well, so it's found more easily by other people.

adschm commented:
Ping.

figgyc commented:
Hi - I think I've made all the changes that you requested now.
I've gotten started on submitting my kernel patch to upstream, however it turns out that there might be some issues with some nonstandard manufacturer IDs (http://lists.infradead.org/pipermail/linux-mtd/2021-February/085068.html)
Am going to start debugging that now - in the meantime not sure if it would be the best idea to merge this or not.

adschm commented on target/linux/ramips/mt76x8/base-files/etc/board.d/02_network:
```diff
@@ -228,6 +229,7 @@ ramips_setup_macs()
 	tplink,archer-c50-v4)
 		wan_mac=$(macaddr_add "$(mtd_get_mac_binary rom 0xf100)" 1)
 		;;
+	creality,wb-01|\
```
Please move that block based on alphabetic sorting.

figgyc commented:
Hi sorry for the long inactivity. Should have fixed all the problems flagged up. 
See https://lkml.org/lkml/2021/2/13/188 I think is the last thing I said about the kernel patch and I havent gotten any replies about it, but from my point of view I think it should be fine to merge? Let me know what you think.

ihrapsa commented:
Ping.

musztarda commented:
ping

fanoush commented:
+1
I'm also waiting for merging this so I can just use imagebuilder from snapshot instead of building from source, please don't sit on this too long.
Or can anyone publish flashable image or openwrt snapshot with imagebuilder pointing to right packages with these changes? Have one WB-01 here waiting to test this. Thanks.

PeterRob commented:
> +1
> I'm also waiting for merging this so I can just use imagebuilder from snapshot instead of building from source, please don't sit on this too long.
> Or can anyone publish flashable image or openwrt snapshot with imagebuilder pointing to right packages with these changes? Have one WB-01 here waiting to test this. Thanks.

You can download from here https://github.com/ihrapsa/KlipperWrt however it it would be very good to get an official build.

ihrapsa commented:
> > +1
> > I'm also waiting for merging this so I can just use imagebuilder from snapshot instead of building from source, please don't sit on this too long.
> > Or can anyone publish flashable image or openwrt snapshot with imagebuilder pointing to right packages with these changes? Have one WB-01 here waiting to test this. Thanks.
> 
> You can download from here https://github.com/ihrapsa/KlipperWrt however it it would be very good to get an official build.

Yup. An official support would facilitate lots of things. I'm still looking forward to it.

wireddown commented:
I have been using this branch successfully since April 28th. Can I do some spot checks or other tests that would help increase yall's confidence in accepting this PR?

```raw
$ git log --oneline -n5
e07d4c3be6 (HEAD -> wb01) Merge branch 'wb01' of https://github.com/figgyc/openwrt into wb01
dfb8ccddcc (origin/wb01) ramips: add support for Creality WB-01
bac20a1834 kernel: add support for Boya BY25Q128AS SPI flash
f4473baf6e ath79: add missing kernel config symbol
e8a5670122 dnsmasq: add ignore hosts dir to dnsmasq init script
```

ihrapsa commented:
Any help reviewing the PR would be much appreciated 🙏🏼

wireddown commented:
> Any help reviewing the PR would be much appreciated 🙏🏼

It looks like this PR is stuck at the [upstream kernel changes](https://lkml.org/lkml/2021/2/13/188) because there's no confident way to correctly identify the SPI memory chip on the WB-01. The first SPI standard only allocated 126 manufacturer IDs (often called the "first bank" of IDs), but there are now many more than that, and it's possible for one ID to match multiple different chips 😱

I haven't done any embedded work with SPI devices, but my understanding from the LKML thread is that the kernel is missing a single mechanism to identify and distinguish SPI devices between the first bank of 126 vendors and the many others who have joined since.

Further, the problem has other factors contributing to its difficulty:

- The device manufacturers _themselves_ are incorrectly implementing the standards for device identification.
   ```diff
     Boya BY25Q128AS responses to SPI command 0x90: "manufacturer and device ID"
   
   + expected: 0x7f 0x7f 0x7f 0x7f 0x7f 0x7f 0x7f 0x7f 0x68
   - actual:                       0x00 0x00 0x00 0x00 0x68 0x17
   ```
- The other standard for identifying SPI devices -- Serial Flash Discoverable Parameter (SFDP) -- is optional and only implemented by some chip makers. Unsurprisingly, the Boya chip on the WB-01 does not implement this.

We need a SPI device expert who can find a method (or knows of another that we don't) to identify _any_ SPI chip using the mixed and partial support we've observed so far.

ihrapsa commented:
Thanks @wireddown for the insight!
 Apparently the guys from flashrom have issues with it as well: this flashrom/flashrom#161 and this flashrom/flashrom#210
We need someone brave and knowledgeable enough to troubleshoot this. :persevere:

adschm commented on target/linux/ramips/dts/mt7628an_creality_wb-01.dts:
```diff
@@ -0,0 +1,114 @@
+			gpios = <&gpio 38 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+
```
Unneeded empty line.

adschm commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -6,6 +6,26 @@ include ./common-tp-link.mk
+  echo 'mtd_write -o 0 -l $$kernel_size write $$file Kernel' \
+    >> $@.install.sh
+  echo 'mtd_write -r -o $$kernel_size -l $$fs_size write $$file RootFS' \
+    >> $@.install.sh
```
I wonder whether this can be written at once with some <-EOF syntax ...

figgyc replied:
I don't think so? It seems as though you have to use `.ONESHELL:` to do heredoc syntax in make but that caused errors like `/usr/bin/env bash: No such file or directory` for me when I tried, and it might also cause problems somewhere else because of how that directive works. There might be a better way but I can't think of one.

adschm commented:
Kernel 5.10 has arrived, so the patch should be applied there as well.

arrowcircle commented:
Hey! Any news on that?

figgyc commented:
Hi - sorry for the inactivity lately, I've been quite busy with my not-software development life stuff. 

I've mostly been waiting on the outcome of the aforementioned patch series discussion upstream. [v3 of the flash collision handling series](https://lore.kernel.org/linux-mtd/20211106113657.172690-1-tudor.ambarus@microchip.com/T/#m4df27e283810ffeae5673731af1ae654d8b2c7d8) was posted recently, although I reckon it will still be a while until it is finished. 

I'll run a rebase on this soon - I did notice though, it seems like [a patch for the BoHong chip of the same ID](https://github.com/openwrt/openwrt/blob/51b61fd57059adeb10a3977e00772af2b76dc40b/target/linux/ramips/patches-5.10/405-mtd-spi-nor-Add-support-for-BoHong-bh25q128as.patch) ended up in the openwrt tree. Not quite sure if/how to deal with that yet, At the very least having two flash patches doing essentially the same thing would probably cause issues, so I'll need to change something before my next rebase. But backporting the flash collisions patch (and presumably the [large set of refactoring patches on which it depends](https://lore.kernel.org/all/20211029172633.886453-19-tudor.ambarus@microchip.com/T/), and probably a lot of other changes in the mtd module since 5.10) would also be quite the endeavour, and I'm not sure how necessary it is in this case.

In this particular scenario I'm actually fairly sure that the collisions module doesn't make much difference, since the BY and BH chips behave seemingly identically, their datasheets are clearly copy pastes of one another, and there is no way to distinguish them at runtime (that we're aware of), so realistically there isn't much change in functionality of porting those patches as opposed to just using the simple patch used currently.  (That's not to say I don't support the patchset, there are other flash collisions covered there which can be distinguished eg some Micron flashes of the same ID with and without SFDP, so it is a good idea to do it properly in the long term.)

I am a bit conflicted on what to do as I am of course enthusiastic about adding support to this device in OpenWrt (it's been my 3D printers' webcam server running stable for several months - wow I started working on this in *January*? Time flies.) but I don't want to make any mistakes, is all. 

Has anyone else got any thoughts? As a last point, have there been any significant refactors in device tree files that I might want to fix? I think I heard stuff about mac address fetching on the mailing list but haven't particularly been keeping track.

gretel commented:
@figgyc thanks, how about forking-off and provide binaries to allow users to test - and eventually progress in regards to the development? not sure if the openwrt context - this PR - will give that.

figgyc commented:
Hi, got bored and decided to rebase today. Tested it and it still seems to work.

For now I just removed the Boya flash patch to avoid conflict with the BoHong one, as I've mentioned it shouldn't make a difference. If you want [it](https://github.com/figgyc/openwrt/commit/bb9f2777c093d24794ec003568595c76cfe38130) back I've created a wb01-old branch on my fork which is just the old commits pre-rebase. As for upstream, the ID collisions patchset hasn't seen activity since I last posted. There have been a lot of other things going on the linux-mtd mailing list though so I'm not surprised by that. It would probably need a rebase at the very least.

So yeah that's the thing. Sorry for not working on this for a while, I've been busy with other projects. I think this would mostly be mergeable with a bit more testing from others, but if I've missed anything do let me know.

There is only one issue I noticed which is this menu in LuCI:

ihrapsa commented:
> 

Hey, I've tried building this with the BoHong patch and it works. However, I've noticed that the wifi mac address changes randomly on every reboot (I think it's not able to access the mac written on the flash). Are you experiencing the same issue? Switching back to an old build with your BoyaMicro patch restors the original mac address.

fanoush commented:
Thank for continuing effort!
BTW just noticed there is  2.0 version https://store.creality.com/products/creality-wifi-cloud-box-2-0-with-tf-card , looks mostly the same with bluetooth added for twice the price , no firmware mentioning 2.0 device here https://www.crealitycloud.com/software-firmware/box?type=1 anyone knows what is inside?

figgyc committed:
ramips: add support for Creality WB-01

The Creality WB-01 (unhelpfully referred to interchangably as the
"Creality Wifi Box" or simply the "Creality Box") is a device intended
to interface Creality brand 3D printers to a cloud service.

Specifications:

- SoC: MediaTek MT7688AN @ 580 MHz
- Flash: BoyaMicro BY25Q128AS (16 MiB, SPI NOR)
  Seems to be a clone/rebrand of the BoHong BH25Q128AS chip (or vice
  versa), support for which was added in commit 51b61fd57
- RAM: 128 MiB DDR2 (Winbond W971GG6SB-25)
- Peripheral: Genesys Logic GL850G 2 port USB 2.0 hub
- I/O: 1x 10/100 Ethernet port, microSD SD-XC Class 10 slot, 4x LEDs,
  2x USB 2.0 ports, micro USB input (for power only), reset button
- FCC ID: 2AXH6CREALITY-BOX
- UART: test pads: (square on silkscreen) 3V3, TX, RX, GND

LEDs:
40 red    - microSD presence (OpenWrt has no function for this I think)
42 green  - "Cloud service or network configuration indicator"
43 blue   - Ethernet
44 yellow - Wi-Fi

Reset button = 38
SD presence switch = 23

MAC addresses:
factory+0x04 wlan and label MAC
factory+0x28 eth0
factory+0x2e has fc:ee:e6:xx:xx:xx here, but it appears to be unused

Stock firmware /proc/mtd
dev:    size   erasesize  name
mtd0: 01000000 00010000 "ALL"
mtd1: 00030000 00010000 "Bootloader"
mtd2: 00010000 00010000 "Config"
mtd3: 00010000 00010000 "Factory"
mtd4: 000d0000 00010000 "Kernel"
mtd5: 00ee0000 00010000 "RootFS"

Flashing:
1) Rename factory.bin to cxsw_update.tar.bz2
2) Copy it to the root of a FAT32 formatted microSD card.
3) Turn on the device, wait for it to start, then insert the card.
The stock firmware reads the install.sh script from this archive, the
build script I added creates one that works in a similar way.

Stock firmware can be downloaded from here:
http://file2-cdn.creality.com/model/cfg/box/latest/cxsw_update.tar.bz2
(source: https://forum.creality.com/topic/1701/ )
Stock firmware supports telnet root@10.10.10.254, password cxswprin

Recovery:
1) Add a file called root_uImage to the root of a FAT32 formatted
   USB drive (tip: a microSD to USB adapter works, as included with
   many Creality printers)
   You can find root_uImage in the stock firmware or you can rename
   sysupgrade.bin.
2) Hold the reset pinhole switch and plug in the device.
3) Keep holding reset for 5 seconds, then release.
4) Wait for the lights to stop flashing.

Signed-off-by: George Brooke <figgyc@figgyc.uk>

drone777 reviewed (approved):
Is this request only approvals missing?

TheRealGramdalf commented:
Heya @figgyc, @adschm (and @ihrapsa if you're still active)
 
I've been working on https://github.com/TheRealGramdalf/KlipperWrt-EX recently, the next evolution of KlipperWRT (by ihrapsa) - as it seems to be mostly unmaintained at the moment. 

It would be incredibly helpful to have this available in the image builder, particularly in docker if possible - it would make the eventual install guide simpler and easier to follow.
I have a creality wifi box, generously donated by Bronius on discord, and I'd be happy to do some testing if needed.

If this won't be getting merged upstream anytime soon, instructions from figgyc on how to build would be great - i.e. what distro did you use to compile, what packages, etc. I've had some weird issues when attempting to build in a debian 11 VM, mostly related to `clang` and `llvm`


Consider this a +1 for the merge :)

chepo92 commented:
Why this is still not merged?

fanoush commented:
> Why this is still not merged?

It is shown right under last post
```
Review required
At least 6 approving reviews are required by reviewers with write access. 
```
From 2 year history of this PR it looks like quite unrealistic requirement :-(
I checked other PRs that got merged into this project recently and did not see so strict requirements anywhere.
However when I sorted PRs by age from oldest https://github.com/openwrt/openwrt/pulls?q=is%3Apr+is%3Aopen+sort%3Acreated-asc  I see many PRs adding hardware support dragging for years like this :-(

I think @figgyc did excellent work fixing code review issues and rebasing it over time and the PR is not really so big and the one controversial issue with SPI flash was resolved long time ago so it is frustrating to see such contributions to be treated like this.

@adschm you made the most competent and thorough review, can you comment what can be realistically done to finally merge this?

motolav commented:
@fanoush The reviewer requirement is just to block people from accidentally merging using github. OpenWrt's main repo hosted by the project elsewhere, this is just a mirror for other people to contribute through github. Maintainers are all volunteers and PRs that appear stalled tend to get hidden here. It could've been made a bit more clear that the PR was complete and ready for review.

chepo92 commented:
How do we get this and others PRs reviewed and aproved for merge?

hauke commented:
Is this device still working with the changes from this pull request?
The change adding support for the SPI NOR flash was removed. Does Linux detect a BoHong bh25q128as SPI NOR chip and is this working fine?

Maybe both companies are selling the same chip and did not change the ID, but just what is printed on the chip.

andytimofeev commented:
it is pity this PR was buried.
Could it be reviewed?

shivajiva101 commented:
@adschm 
I maintain the working fork of KlipperWRT [here](https://github.com/shivajiva101/KlipperWrt)
It's fair and accurate to say this pr is stable and has no issues purely based on the success of KlipperWRT itself. That is until v24 releases where mmc breaks for the device. Possibly related to #17446 not encompassing this hardware design

TheRealGramdalf commented:
@shivajiva101 I have some ideas and works in progress that would make klipperwrt much easier to use and maintain, and I was planning on contributing when I had some spare time. One of the main blockers is in fact that this is an unofficial build, so getting this merged would be super helpful.
If @adschm is still available for review I'm happy to help with testing when it comes to v24, as I have a wb-01 that I can flash without much hassle. I can also attest to it running fine; I haven't done an in depth inspection but it works perfectly for what I need it for and everything I've interacted with so far works as expected.

chepo92 commented:
Oh come on guys 6 approving reviews needed to merge ?, there are lot of others PRs which add hw support that have been merged without that, and this PR has been more than tested

chepo92 commented:
Anyway we have 3 similar attempts, I think the lastest and more complete is https://github.com/openwrt/openwrt/pull/15021
